datafusion/physical-expr/src/aggregate/percentile.rs

```diff
@@ -0,0 +1,712 @@
+        Ok(Field::new(&self.name, self.data_type.clone(), true))
+    }
+
+    fn create_accumulator(&self) -> Result<Box<dyn Accumulator>> {
+        Ok(Box::new(PercentileAccumulator {
+            percentile_score: 0.5,
+            interpolation_type: PercentileInterpolationType::Continuous,
+            data_type: self.data_type.clone(),
+            arrays: vec![],
+            all_values: vec![],
+        }))
+    }
+
+    fn state_fields(&self) -> Result<Vec<Field>> {
+        //Intermediate state is a list of the elements we have collected so far
+        let field = Field::new("item", self.data_type.clone(), true);
+        let data_type = DataType::List(Arc::new(field));
+
+        Ok(vec![Field::new(
+            format_state_name(&self.name, "median"),
+            data_type,
+            true,
+        )])
+    }
+
+    fn expressions(&self) -> Vec<Arc<dyn PhysicalExpr>> {
+        vec![self.expr.clone()]
+    }
+
+    fn name(&self) -> &str {
+        &self.name
+    }
+}
+
+impl PartialEq<dyn Any> for Median {
+    fn eq(&self, other: &dyn Any) -> bool {
+        down_cast_any_ref(other)
+            .downcast_ref::<Self>()
+            .map(|x| {
+                self.name == x.name
+                    && self.data_type == x.data_type
+                    && self.expr.eq(&x.expr)
+            })
+            .unwrap_or(false)
+    }
+}
+
+#[derive(Debug)]
+/// The accumulator for median/quantile_cont/quantile/disc aggregate functions
+/// It accumulates the raw input values as `ScalarValue`s
+///
+/// The intermediate state is represented as a List of scalar values updated by
+/// `merge_batch` and a `Vec` of `ArrayRef` that are converted to scalar values
+/// in the final evaluation step so that we avoid expensive conversions and
+/// allocations during `update_batch`.
+struct PercentileAccumulator {
+    percentile_score: f64,
+    interpolation_type: PercentileInterpolationType,
+    data_type: DataType,
+    arrays: Vec<ArrayRef>,
+    all_values: Vec<ScalarValue>,
+}
+
+macro_rules! safe_average {
+    (f32, $v1:expr, $v2:expr) => {
+        $v1 / 2.0 + $v2 / 2.0
+    };
+    ( f64, $v1:expr, $v2:expr) => {
+        $v1 / 2.0 + $v2 / 2.0
+    };
+    ($val_type:ty, $v1:expr, $v2:expr) => {
+        match $v1.checked_add($v2) {
+            Some(sum) => sum / (2 as $val_type),
+            None => $v1 / (2 as $val_type) + $v2 / (2 as $val_type),
+        }
+    };
+}
+
+// Example: `target_percentile` is 0.12 and it's landed between dp1 and dp2
+// dp1 has percentile 0.10 and value 0
+// dp2 has percentile 0.20 and value 100
+// `quantile_cont()` do linear interpolation:
+//      Then interpolation result = 0 + (0.12 - 0.10) / (0.20 - 0.10) * (100 - 0)
+//                                = 20
+// `quantile_disc()` choose the closer dp (pick one with lower percentile if equally close)
+//      `target_percentile` is closer to dp1's percentile, result = 0
+macro_rules! interpolate_logic {
```

Comment:
This complex macro is used 1. Reduce float arithmetic errors 2. Avoid potential overflows
Looks like can be avoided if multiply and divide operation on `ScalarValue`s like this one is available
https://github.com/apache/arrow-datafusion/blob/672f5bdfe4df2270f5ad7b1a49bb1135acd5f9e7/datafusion/common/src/scalar.rs#L2057-L2060C4
seem they're available in `arrow-rs`? But require to add them into `datafusion`